src/ui/public/state_management/unhash_states.js

```diff
@@ -0,0 +1,46 @@
+import { parse as parseUrl, format as formatUrl } from 'url';
+import { mapValues } from 'lodash';
+
+export function UnhashStatesProvider(getAppState, globalState) {
+  const getDefaultStates = () => [getAppState(), globalState].filter(Boolean);
+
+  this.inAbsUrl = (url, states = getDefaultStates()) => {
+    if (!url) return url;
+
+    const entireUrlParsed = parseUrl(url, true);
+    if (!entireUrlParsed.hostname) {
+      // passing a url like "localhost:5601" or "/app/kibana" should be prevented
+      throw new TypeError(
+        'Only absolute urls should be passed to `unhashStates.inAbsUrl()`. ' +
+        'Unable to detect url hostname.'
+      );
+    }
+
+    if (!entireUrlParsed.hash) return url;
+
+    const appUrl = entireUrlParsed.hash.slice(1); // trim the #
+    if (!appUrl) return url;
+
+    const appUrlParsed = parseUrl(entireUrlParsed.hash.slice(1), true);
+    if (!appUrlParsed.query) return url;
+
+    const unhashedAppQuery = this.inObject(appUrlParsed.query || {}, states);
+    return formatUrl({
+      protocol: entireUrlParsed.protocol,
+      host: entireUrlParsed.host,
+      pathname: entireUrlParsed.pathname,
+      query: entireUrlParsed.query,
+      hash: formatUrl({
+        pathname: appUrlParsed.pathname,
+        query: unhashedAppQuery,
+      })
+    });
+  };
+
+  this.inObject = (query, states = getDefaultStates()) => {
+    return mapValues(query, (val, key) => {
+      const state = states.find(s => key === s.getQueryParamName());
+      return state ? state.toRISON() : val;
+    });
+  };
+}
```

Comment:
The more I go down this path the more confusing it seems to call things hashes everywhere except here.